app/src/main/java/org/listenbrainz/android/data/repository/ArtistRepositoryImpl.kt

```diff
@@ -34,26 +38,51 @@ class ArtistRepositoryImpl @Inject constructor(
                 }
             }
 
-    override suspend fun addArtists(): Boolean {
-        val artists = AlbumsData().fetchAlbums()
+    override suspend fun addArtists(userRequestedRefresh: Boolean): Boolean {
+        val artists = AlbumsData.fetchAlbums(userRequestedRefresh)
             .map {
                 it.toArtistEntity()
             }
             .distinct()
-        for (artist in artists) {
-            artist.songs.addAll(addAllSongsOfArtist(artist.toArtist()).map {
-                it.toSongEntity()
-            })
-            artist.albums.addAll(addAllAlbumsOfArtist(artist.toArtist()).map {
-                it.toAlbumEntity()
-            })
+        
+        // Deleting artists that are not present in the device anymore.
+        artistDao.getArtistEntitiesAsList().forEach { artistEntity ->
+            if (!artists.contains(artistEntity))
+                artistDao.deleteArtist(artistEntity.name)
         }
+        
+        lateinit var songsJob : Deferred<Unit>
+        lateinit var albumsJob : Deferred<Unit>
+        
+        withContext(Dispatchers.IO){
+            // Both jobs are being executed simultaneously.
+            songsJob = async {
+                for (artist in artists) {
+                    // Here, if userRequestedRefresh is true, it will refresh songs cache which is what we expect from refreshing
+                    artist.songs.addAll(addAllSongsOfArtist(artist.toArtist(), userRequestedRefresh).map {
+                        it.toSongEntity()
+                    })
+                }
+            }
+            albumsJob = async {
+                for (artist in artists) {
+                    // We do not need to refresh cache (songsListCache) here as it already got refreshed above when we created list of albums.
+                    artist.albums.addAll(addAllAlbumsOfArtist(artist.toArtist()).map {
+                        it.toAlbumEntity()
+                    })
+                }
+            }
+        }
+        
+        songsJob.await()
```

Comment:
We should write callback functions rather than blocking this function in IO to proceed.

Reply:
Blocking has to be done in order for the artist lists to be ready for addition. The main motive here is that we are executing both the functions above asynchronously rather than doing one job and then other which is less efficient.

Correct me if I misunderstood.